zerver/tests/test_bots.py

```diff
@@ -1524,6 +1524,47 @@ def test_patch_bot_role(self) -> None:
         result = self.client_patch(f"/json/bots/{user_profile.id}", req)
         self.assert_json_error(result, "Must be an organization administrator")
 
+    def test_patch_bot_type(self) -> None:
+        self.login("hamlet")
+        bot_info = {
+            "full_name": "The Bot of Hamlet",
+            "short_name": "hambot",
+        }
+        result = self.client_post("/json/bots", bot_info)
+        self.assert_json_success(result)
+
+        email = "hambot-bot@zulip.testserver"
+        bot_profile = self.get_bot_user(email)
+
+        req = dict(bot_type=UserProfile.INCOMING_WEBHOOK_BOT)
+        result = self.client_patch(f"/json/bots/{bot_profile.id}", req)
+        self.assert_json_success(result)
+
+        req = dict(bot_type=UserProfile.OUTGOING_WEBHOOK_BOT)
+        result = self.client_patch(f"/json/bots/{bot_profile.id}", req)
+        self.assert_json_error(result, "Missing service_payload_url.")
+
+        req_outgoing = dict(
+            bot_type=UserProfile.OUTGOING_WEBHOOK_BOT,
+            service_payload_url=orjson.dumps("http://foo.bar.com").decode(),
+            service_interface=Service.GENERIC,
+        )
+        result = self.client_patch(f"/json/bots/{bot_profile.id}", req_outgoing)
+        self.assert_json_success(result)
+
+        req = dict(bot_type=UserProfile.EMBEDDED_BOT)
+        result = self.client_patch(f"/json/bots/{bot_profile.id}", req)
+        self.assert_json_success(result)
```

Comment:
Whenever you have an `assert_json_success`, you need to also be checking that the change actually took effect properly. A good technique would be to `GET` the bot user and inspect the properties of what was returned.